### DISCUSSION TITLE
Issues when running Eclipse on GTK3

### DISCUSSION THREAD

**Discussion by wimjongman** [eclipse-platform/eclipse.platform.swt#375] [Q&A] [answered]
I have an RCP that a user is running on GTK3. I also tried it, and it bombs out on various locations (logs attached). I must be doing something wrong. Can anyone help me?

org.eclipse.swt.internal.gtk.version=3.24.20
org.eclipse.swt.internal.webkitgtk.version=2.36.6

E.g., I have this code:

```
mainContainer = new Composite(managedForm.getForm().getBody(), NONE);
mainContainer.setBackground(display.getSystemColor(SWT.COLOR_WHITE));
```

Which ends in error:

```
!ENTRY org.eclipse.ui 4 0 2022-09-07 01:49:56.591
!MESSAGE Unhandled event loop exception
!STACK 0
java.lang.IllegalArgumentException: Argument not valid
	at org.eclipse.swt.SWT.error(SWT.java:4899)
	at org.eclipse.swt.SWT.error(SWT.java:4833)
	at org.eclipse.swt.SWT.error(SWT.java:4804)
	at org.eclipse.swt.widgets.Widget.error(Widget.java:565)
	at org.eclipse.swt.widgets.Control._setBackground(Control.java:5186)
	at org.eclipse.swt.widgets.Control.setBackground(Control.java:5177)
	at com.remainsoftware.common.licensing.ui.AbstractLicenseComposite._createControl(AbstractLicenseComposite.java:217)
	at com.remainsoftware.common.licensing.ui.AbstractLicenseComposite.lambda$1(AbstractLicenseComposite.java:200)
```

[config.txt](https://github.com/eclipse-platform/eclipse.platform.swt/files/9501652/config.txt)
[errorlog.log](https://github.com/eclipse-platform/eclipse.platform.swt/files/9501653/errorlog.log)

**Comment by akurtakov** [comment]
Can you share standalone snippet reproducer?

**Reply by wimjongman** [reply]
I don't know how. 

From your reaction, I see that it is not something that is common. I will run a standalone Eclipse then install my software and do a compare to see any weirdness in bundles. I probably pull in something that confuses SWT.

**Comment by wimjongman** [comment, marked as answer]
Ok, I found it. It turns out that system colors can be disposed in gtk but not in win. 

The following snippet runs on windows but not on ubuntu:

```
import org.eclipse.swt.SWT;
import org.eclipse.swt.graphics.Color;
import org.eclipse.swt.layout.GridLayout;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.Shell;

public class Snippet {

	public static void main(String[] args) {
		Display display = new Display();
		Shell shell = new Shell(display);
		shell.setLayout(new GridLayout(1, false));
		Color systemColor = display.getSystemColor(SWT.COLOR_WHITE);
		shell.setBackground(systemColor);
		systemColor.dispose();
		systemColor = display.getSystemColor(SWT.COLOR_WHITE);
		shell.setBackground(systemColor);

		shell.pack();
		shell.open();
		while (!shell.isDisposed()) {
			if (!display.readAndDispatch())
				display.sleep();
		}
		display.dispose();
	}
}
```

**Comment by akurtakov** [comment]
Is this using latest version of SWT ?

**Reply by wimjongman** [reply]
For this test I used 4.24.0. Do you want me to test this using the ibuilds?

Version: 2022-06 (4.24.0)
Build id: 20220609-1112

\> ss swt
org.eclipse.e4.ui.css.swt_0.14.500.v20220511-1639
org.eclipse.e4.ui.css.swt.theme_0.13.100.v20220310-1056
org.eclipse.e4.ui.swt.gtk_1.1.100.v20210108-1832
org.eclipse.e4.ui.workbench.addons.swt_1.4.400.v20211102-0453
org.eclipse.e4.ui.workbench.renderers.swt_0.15.500.v20220511-1638
org.eclipse.e4.ui.workbench.swt_0.16.500.v20220506-1520
org.eclipse.swt_3.120.0.v20220530-1036
org.eclipse.swt.gtk.linux.x86_64_3.120.0.v20220530-1036
